asab/library/service.py

```diff
@@ -375,45 +375,134 @@ async def _list(self, path, providers):
 		return items
 
 
-	async def _read_disabled(self):
-		# `.disabled.yaml` is read from the first configured library
-		# It is applied on all libraries in the configuration.
-		disabled = await self.Libraries[0].read('/.disabled.yaml')
+	async def _read_disabled(self, publish_changes=False):
+		old_disabled = self.Disabled.copy()
+		old_disabled_paths = list(self.DisabledPaths)
 
-		if disabled is None:
-			self.Disabled = {}
-			self.DisabledPaths = []
-			return
+		# Read the file
+		disabled_file = await self.Libraries[0].read('/.disabled.yaml')
 
-		try:
-			disabled = yaml.load(disabled, Loader=yaml.CSafeLoader)
-		except Exception:
+		if disabled_file is None:
 			self.Disabled = {}
 			self.DisabledPaths = []
-			L.exception("Failed to parse '/.disabled.yaml'")
-			return
+		else:
+			try:
+				disabled_data = yaml.load(disabled_file, Loader=yaml.CSafeLoader)
+			except Exception:
+				L.exception("Failed to parse '/.disabled.yaml'")
+				self.Disabled = {}
+				self.DisabledPaths = []
+				return
+
+			if disabled_data is None:
+				self.Disabled = {}
+				self.DisabledPaths = []
+				return
+
+			if isinstance(disabled_data, set):
+				# Backward compatibility (August 2023)
+				self.Disabled = {key: '*' for key in disabled_data}
+				self.DisabledPaths = []
+			else:
+				self.Disabled = {}
+				self.DisabledPaths = []
+				for k, v in disabled_data.items():
+					if k.endswith('/'):
+						self.DisabledPaths.append((k, v))
+					else:
+						self.Disabled[k] = v
 
-		if disabled is None:
-			self.Disabled = {}
-			self.DisabledPaths = []
+		self.DisabledPaths.sort(key=lambda x: len(x[0]))
+
+		# If requested, compare old and new disables to notify subscribers via Library.change!
+		if publish_changes:
+			await self._publish_change_for_disabled_diff(old_disabled, old_disabled_paths)
+
+	async def _publish_change_for_disabled_diff(self, old_disabled, old_disabled_paths):
+		"""
+		Compare old and new disabled data and publish Library.change!
+		if any subscribed path is affected for the specific subscription target.
+		"""
+
+		if not self.Libraries:
 			return
 
-		if isinstance(disabled, set):
-			# This is for a backward compatibility (Aug 2023)
-			self.Disabled = {key: '*' for key in self.Disabled}
-			self.DisabledPaths = []
+		provider = self.Libraries[0]
```

Comment:
Please add a comment that the disabled file is read only from the topmost (0) provider

Reply:
this should be also somewhere in the docs btw

Reply:
@PremyslCerny By ASAB’s design the first provider is the single source-of-truth for all disable rules (/.disabled.yaml), and only its subscriptions drive the disable logic and change notifications.